enhancements/builds/multi-arch-builds.md

```diff
@@ -0,0 +1,456 @@
+
+### Goals
+
+* Provide users of Builds with a minimal — in cases where builds are based on images we provide, preferably zero — effort path to having their images ready to run on any node in their multi-architecture clusters.
+
+### Non-Goals
+
+* Building images for architectures for which the cluster has no running nodes.
+* Building multi-arch images which depend on base images which are not multi-arch.
+
+## Proposal
+
+### openshift-controller-manager
+
+Instead of creating one (1) pod to perform a build followed by an optional
+push, create N+1 pods when building for N architectures.
+
+The first N pods will build an ordinary image for their respective targets and,
+if the build spec specifies an output location, push their images and update a
+platform-specific build status field in the build resource.
+
+The last pod will use functionality to be added to the openshift builder to
+build a manifest list and push it to the output location.
+
+### openshift-builder
+
+The builder binary is currently invoked by multiple names to perform multiple
+tasks in build pods.
+
+We'll extend openshift-builder's `openshift-docker-build` and
+`openshift-sti-build` commands to perform the build as they already do, with
+two exceptions: if the build spec specifies a target platform, the build will
+target that platform for the build, and the build status will be written to the
+corresponding platform-specific status field in the Build object which owns the
+build pod.
+
+We'll extend openshift-builder by adding another command it can be invoked by
+which will tell it to monitor the status of the other builder pods. If the
+builder pods all complete successfully, and the build spec specifies an output
+location, it will build and push a manifest list using the information gathered
+from the build status information. As its final act, it will attempt to untag
+the individual images from the registry.
+
+Note: untagging should be friendlier to registries which perform garbage
+collection, particularly for cases where the build fails on one or more
+architectures, and as a result a manifest list is not pushed. It should also
+reduce "noise" in the list of tags reported by registries which support listing
+the tags in a repository.
+
+### Workflow Description
+
+#### Variation 1/3: SCM integrated
+1. Application developer merges a PR to their repository's main branch.
+1. The SCM platform notifies OpenShift via a webhook.
+1. OpenShift has previously associated the webhook with triggering a BuildConfig.
+
+#### Variation 2/3: not SCM integrated
+1. Application developer manually triggers a BuildConfig.
+
+#### Variation 3/3: not SCM integrated
+1. Application developer creates a Build.
+
+#### Common
+1. The openshift-controller-manager starts N+1 build pods.
+1. Platform-specific images are pushed to the registry.
+1. The build completes, and a manifest list is pushed to the registry.
+1. Platform-specific images are untagged from the registry.
```

Comment:
Is this tagging/untagging due to some restriction on the Build api?

Reply:
Tagging is a hedge against aggressive garbage collection of untagged images at the registry.
The untagging is an attempt at being tidy and not littering the set of tags in the repository, but I'm not all that attached to that.

Reply:
> If there is a bool and it is set to false, then I think we shouldn't clean up these tags.

if there is a bool and it's set to false, we won't be building multiarch+manifestlists at all, we'll be building a single image and pushing it as a manifest.

Reply:
to address @arewm's use case we'd need a "don't remove tags" option which would ensure that the arch specific images remained tagged on the registry, in addition to pushing the manifestlist tag.

a not unreasonable suggestion, though i've also yet to see how the arch specific tag names are generated/specified (needs to be answered regardless of whether we allow the tags to be left in place or always remove them)

Reply:
I am likely following comments out of order, but if a bool exists, it wouldn't be a bool to indicate whether a manifestlist would be produced _at all_? This doesn't seem to be the thought based on the above comment.

If a manifestlist is always generated, then the option to keep intermediate tags would enable the non-manifest list behavior if desired even if images are still included as part of a manifestlist.

Reply:
I think we are on the same page. I was just confused by the phrase

> to address @arewm's use case we'd need a "don't remove tags" option which would ensure that the arch specific images remained tagged on the registry, in addition to pushing the manifestlist tag.

and took it to override your initial comment/suggestion to the point of changing what the boolean was doing.